packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -180,12 +180,73 @@ class LineBreak extends TextBoundary {
   }
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminator exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    return TextPosition(
+      offset: getTextBoundaryAt(position).start,
+    );
+  }
+
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    return TextPosition(
+      offset: getTextBoundaryAt(position).end,
+      affinity: TextAffinity.upstream,
+    );
+  }
+
+  // Returns the [TextRange] representing a paragraph that bounds the given
+  // `textPosition`. The `textPosition` is bounded by either a line terminator
+  // in each direction of the text, or if there is no line terminator in a given
+  // direction then the bound extends to the start/end of the document in that
+  // direction. The returning range includes the line terminator.
+  @override
+  TextRange getTextBoundaryAt(TextPosition position) {
+    final Iterator<int> codeUnitIter = _text.codeUnits.iterator;
+    final int tappedTextOffset = position.offset;
```

Comment:
Nit: Maybe make the language here agnostic of "tap", since this method doesn't necessarily have to do with tapping on the face of it.